man/rbldns.8
.TH djbdnscurve6: rbldns 8
.SH NAME
rbldns \- DNSCurve enabled IP-address-listing DNS server
.SH DESCRIPTION
.B rbldns
is an IP-address-listing DNS server.
It accepts iterative DNS queries
from hosts around the Internet asking about various IP addresses.
It provides responses showing whether the addresses
are on a locally configured list,
such as RBL or DUL.
Encrypted DNS queries in the 
.I DNSCurve
format are honored and replied encrypted als well,
given, the name server's
.I public key
is generated by means of
.B dnscurve-keygen
and used as primary DNS name for
.IR rbldns .
.SH CONFIGURATION
Normally 
.B rbldns
is set up by the
.B rbldns-conf
program.

.B rbldns
runs chrooted in the directory
specified by the 
.I $ROOT
environment variable,
under the uid and gid
specified by the 
.I $UID
and 
.I $GID
environment variables.

.B rbldns
listens for incoming UDP packets
addressed to port 53 of 
.IR $IP .
It does not listen for TCP queries.
Specifying
.I 0.0.0.0 
or 
.I ::
results in listing to all available 
IP addresses and interfaces (for IPv6)
respectively. In case
.I $IP
is specified as the pseudo IP address
.IR :0 ,
.B rbldns
is forced to bind to all available IPv4 and IPv6
addresses simultaneously.


.B rbldns
handles IPv4 domains of the form
.IR d.c.b.a\fR.\fIbase
where 
.I a.b.c.d
is the IPv4 address in question
and 
.I base
is given by the 
.I $BASE
environment variable.

.B rbldns
handles IPv6 domains of the form
.IR z.y.x...d.c.b.a\fR.\fIbase
where 
.IR z.y.x...d.c.b.a\fR.\fIbase
is the IPv6 address in question
and 
.I base
is given by the 
.I $BASE
environment variable.
While particular host addresses can be given,
only the relevant 64 routing bits can be specified.

.B rbldns
rejects
inverse queries, non-Internet-class queries, truncated packets,
packets that contain anything other than a single query,
query types other than A, TXT, or *,
and queries for domains outside 
.IR $BASE .
It looks up other queries inside 
.IR data.cdb ,
a binary file created by
.BR rbldns-data .
.SH SEE ALSO
curvedns-keygen(8),
rbldns-conf(8),
rbldns-data(8)
.SH REFERENCE
https://cr.yp.to/djbdns.html


src/rbldns.c
#include <unistd.h>
#include "str.h"
#include "byte.h"
#include "ip.h"
#include "open.h"
#include "env.h"
#include "cdbread.h"
#include "dns.h"
#include "dd.h"
#include "response.h"
#include "logmsg.h"
#include "uint_t.h"
#include "scan.h"

#define WHO "rbldns"
const char *fatal = "rbldns";
int flagcurved;

static char *base;

static struct cdb c;
static char data[100 + IP6_FMT];

static int doit(char *q,char qtype[2])
{
  int flaga;
  int flaga4;
  int flagtxt;
  int flagip6;
  char ch;
  char reverseip4[4];
  char reverseip6[16];
  char ip4[4];	
  char ip5[16]; /* aehm */
  char ip6[16];
  uint32 ipnum4;
  uint64 ipnum5; /* aehm */
  struct uint128_t ipnum6;
  uint32 dlen;
  int r;
  int i;
  stralloc ipstring = {0};
  stralloc tmp = {0};

  flaga = byte_equal(qtype,2,DNS_T_A);
  flaga4 = byte_equal(qtype,2,DNS_T_AAAA);
  flagtxt = byte_equal(qtype,2,DNS_T_TXT);
  if (byte_equal(qtype,2,DNS_T_ANY)) flaga = flaga4 = flagtxt = 1;
  if (!(flaga || flaga4) && !flagtxt) goto REFUSE;

  if (flaga || flagtxt) {
    r = dd4(q,base,reverseip4);
    if (r < 0) goto IPV6;
    if (r != 4) goto REFUSE;
    uint32_unpack(reverseip4,&ipnum4);
    uint32_pack_big(ip4,ipnum4);
    if (ip4_bytestring(&ipstring,ip4,32)) return 0;

    for (i = 32; i > 0; --i) {
      if (!stralloc_copys(&tmp,"")) return 0;
      if (!stralloc_catb(&tmp,ipstring.s,i)) return 0;
      r = cdb_find(&c,tmp.s,i);
      if (r == -1) return 0;
      if (r) goto BASE;
    }
    if (!r) { response_nxdomain(); return 1; }
  }


  IPV6:
  if (flaga4 || flagtxt) {
    flagip6 = 1;
    if (dd6(q,base,reverseip6) != 16) goto REFUSE;

    uint128_unpack(reverseip6,&ipnum6);	// IPv6 incl. link token 
    uint128_pack_big(ip6,ipnum6);
    if (ip6_bytestring(&ipstring,ip6,128)) return 0;
    if (!stralloc_copys(&tmp,"^")) return 0;
    if (!stralloc_catb(&tmp,ipstring.s,128)) return 0;
    r = cdb_find(&c,tmp.s,129);
    if (r) goto BASE;

    uint64_unpack(reverseip6+8,&ipnum5); // IPv6 net-id only
    uint64_pack_big(ip5,ipnum5);
    if (ip6_bytestring(&ipstring,ip5,64)) return 0;

    for (i = 64; i > 0; --i) {
      if (!stralloc_copys(&tmp,"^")) return 0;
      if (!stralloc_catb(&tmp,ipstring.s,i)) return 0;
      r = cdb_find(&c,tmp.s,i);
      if (r == -1) return 0;
      if (r) goto BASE;
    }
    if (!r) { response_nxdomain(); return 1; }
  } 


  BASE:  
  r = cdb_find(&c,"",0);
  if (r == -1) return 0;
  if (r && ((dlen = cdb_datalen(&c)) >= 4)) {
    if (dlen > 256) dlen = 256;
    if (cdb_read(&c,data,dlen,cdb_datapos(&c)) == -1) return 0;
  }
  else {
    dlen = 12;
    byte_copy(data,dlen,"\177\0\0\2Listed $");
  }

  if ((dlen >= 5) && (data[dlen - 1] == '$')) {
    --dlen;
    if (flagip6) 
      dlen += ip6_fmt(data + dlen,ip6);
    else 
      dlen += ip4_fmt(data + dlen,ip4);
  }

  if (flaga) {
    if (!response_rstart(q,DNS_T_A,2048)) return 0;
    if (!response_addbytes(data,4)) return 0;
    response_rfinish(RESPONSE_ANSWER);
  }
  if (flaga4) {
    if (!response_rstart(q,DNS_T_AAAA,2048)) return 0;
    if (!response_addbytes(data,16)) return 0;
    response_rfinish(RESPONSE_ANSWER);
  }
  if (flagtxt) {
    if (!response_rstart(q,DNS_T_TXT,2048)) return 0;
    ch = dlen - 4;
    if (!response_addbytes(&ch,1)) return 0;
    if (!response_addbytes(data + 4,dlen - 4)) return 0;
    response_rfinish(RESPONSE_ANSWER);
  }

  return 1;


  REFUSE:
  response[2] &= ~4;
  response[3] &= ~15;
  response[3] |= 5;
  return 1;
}

int respond(char *q,char qtype[2],char ip[16])
{
  int fd;
  int result;

  fd = open_read("data.cdb");
  if (fd == -1) return 0;
  cdb_init(&c,fd);
  result = doit(q,qtype);
  cdb_free(&c);
  close(fd);
  return result;
}

const char *starting = "starting rbldns ";

void init_server(void)
{
  char *x;

  x = env_get("BASE");
  if (!x)
    logmsg(WHO,111,ERROR,"$BASE not set");
  if (dns_domain_fromdot(&base,x,str_len(x)) <= 0)
    logmsg(WHO,111,FATAL,"unable to parse $BASE"); 
}
